src/saga.js

```diff
@@ -0,0 +1,135 @@
+/* @flow */
+
+import { put, select, call, take, cancelled, fork } from 'redux-saga/effects';
+import { eventChannel } from 'redux-saga';
+import { NetInfo } from 'react-native';
+import checkInternetAccess from './checkInternetAccess';
+import { connectionChange } from './actionCreators';
+import type { NetworkState } from './types';
+
+type Arguments = {
+  timeout?: number,
+  pingServerUrl?: string,
+  withExtraHeadRequest?: boolean,
+  checkConnectionInterval?: number,
+};
+
+export default function* networkEventsListenerSaga(
+  {
+    timeout = 3000,
+    pingServerUrl = 'https://google.com',
+    withExtraHeadRequest = true,
+    checkConnectionInterval = 0,
+  }: Arguments = {},
+): Generator<*, *, *> {
+  yield fork(
+    handleNetInfoChangeChannel,
+    timeout,
+    pingServerUrl,
+    withExtraHeadRequest,
+  );
+  if (checkConnectionInterval) {
+    yield fork(
+      handleCheckConnectionIntervalChannel,
+      timeout,
+      pingServerUrl,
+      checkConnectionInterval,
+    );
+  }
+}
+
+function* handleNetInfoChangeChannel(
+  timeout: number,
+  pingServerUrl: string,
+  withExtraHeadRequest: boolean,
+): Generator<*, *, *> {
+  const netInfoChangeChannel = yield call(createNetInfoConnectionChangeChannel);
+
+  function* handleNetInfoChange(isConnected: boolean) {
+    if (!isConnected) {
+      yield call(handleConnectivityChange, isConnected);
+    } else {
+      yield call(checkInternet, timeout, pingServerUrl);
+    }
+  }
+
+  try {
+    while (true) {
+      const isConnected = yield take(netInfoChangeChannel);
+      yield fork(
+        withExtraHeadRequest ? handleNetInfoChange : handleConnectivityChange,
+        isConnected,
+      );
+    }
+  } finally {
+    if (yield cancelled()) {
+      netInfoChangeChannel.close();
+    }
+  }
+}
+
+function* handleCheckConnectionIntervalChannel(
+  timeout?: number,
+  pingServerUrl?: string,
+  checkConnectionInterval: number,
```

Comment:
let's change this param name to just `interval`, since it's implementation detail and it's too redundant.